rotkehlchen/exchanges/coinbase.py

```diff
@@ -285,49 +381,56 @@ def _api_query(
             options: dict[str, Any] | None = None,
             ignore_pagination: bool = False,
     ) -> list[Any]:
-        """Performs a coinbase API Query for endpoint
+        """
+        Performs a Coinbase API Query for endpoint.
 
         You can optionally provide extra arguments to the endpoint via the options argument.
         If you want just the first results then set ignore_pagination to True.
         """
         all_items: list[Any] = []
         had_4xx = False
         request_verb = 'GET'
-        # initialize next_uri before loop
-        next_uri = f'/{self.apiversion}/{endpoint}'
+        next_uri = f'/{self.apiversion}/{endpoint}'  # initialize next_uri before loop
         timeout = CachedSettings().get_timeout_tuple()
+
         if options:
             next_uri += f'?{urlencode(options)}'
         while True:
-            timestamp = str(int(time.time()))
-            message = timestamp + request_verb + next_uri
-
-            signature = hmac.new(
-                self.secret,
-                message.encode(),
-                hashlib.sha256,
-            ).hexdigest()
-            log.debug('Coinbase API query', request_url=next_uri)
-
-            self.session.headers.update({
-                'CB-ACCESS-SIGN': signature,
-                'CB-ACCESS-TIMESTAMP': timestamp,
-            })
+            if self.is_legacy_api_key:
+                timestamp = str(int(time.time()))
+                message = timestamp + request_verb + next_uri
+
+                signature = hmac.new(
+                    self.secret,
+                    message.encode(),
+                    hashlib.sha256,
+                ).hexdigest()
+                log.debug('Coinbase API query', request_url=next_uri)
+
+                self.session.headers.update({
+                    'CB-ACCESS-SIGN': signature,
+                    'CB-ACCESS-TIMESTAMP': timestamp,
+                })
+            else:
+                uri = f'{request_verb} {self.host}/{self.apiversion}/{endpoint}'
+                log.debug('Coinbase CDP API query', request_url=uri)
+                token = self.build_jwt(uri)
+                self.session.headers.update({
+                    'Authorization': f'Bearer {token}',
+                })
 
             full_url = self.base_uri + next_uri
+
             try:
                 response = self.session.get(full_url, timeout=timeout)
             except requests.exceptions.RequestException as e:
                 raise RemoteError(f'Coinbase API request failed due to {e!s}') from e
 
-            if response.status_code in (401, 429) and had_4xx is False:
+            if response.status_code in (401, 403, 429) and had_4xx is False:
```

Comment:
hmm you resolved this: https://github.com/rotki/rotki/pull/7903#discussion_r1593552039 comment without explaining.

Why did you move it here? What's the reason?

Before if we got a 403 (FORBIDDEN) error we raised a `CoinbasePermissionError`. Meaning that the api key does not have permission for the resource they are trying to access.

Now with your change a (403 FORBIDDEN) just retries. That seems like a wrong change. Can you explain the logic ofthe change here?

Reply:
Yeah, makes sense. Fixed the code.